### PR TITLE
COcclusion done, I hope

### PR DESCRIPTION
**Pull request by majesticCoding**

### CODE REVIEW

File: src/entities/Entity.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class CEntity : public CPlaceable
 	bool IsVisible(void) { return m_rwObject && bIsVisible && GetIsOnScreen(); }
 	bool IsVisibleComplex(void) { return m_rwObject && bIsVisible && GetIsOnScreenComplex(); }
 // TODO(MIAMI):
```
Comment (erorcun) [inline src/entities/Entity.h:159]: you can remove the TODO :)

File: src/render/Occlusion.cpp [hunk 2]
```diff
@@ -38,7 +55,408 @@ COcclusion::AddOne(float x, float y, float z, float width, float length, float h
...
+	if (!CalcScreenCoors(pos, &gCenterOnScreen, &outX, &outY) || gCenterOnScreen.z < -150.0f || gCenterOnScreen.z > 300.0f) {
+		return false;
+	}
+
+	occl->radius = Max(width, length) * 0.35f + gCenterOnScreen.z;
+
+	CVector vec[3];
+	vec[0] = { width / 2.0f * sinf(angle*0.000095f), -width / 2.0f * cosf(angle*0.000095f), 0.0f };
+	vec[1] = { length / 2.0f * cosf(angle*0.000095f), length / 2.0f * sinf(angle*0.000095f), 0.0f };
+	vec[2] = { 0.0f, 0.0f, height / 2.0f };
```
Comment (erorcun) [inline src/render/Occlusion.cpp:197]: assigments with braces are C++14 or 19 or something and should be avoided

File: src/entities/Entity.h (hunk 1 above)
Comment (majesticCoding) [reply src/entities/Entity.h]: removed

File: src/render/Occlusion.cpp (hunk 2 above)
Comment (majesticCoding) [reply src/render/Occlusion.cpp]: fixed it

File: src/render/Occlusion.cpp [hunk 3]
```diff
@@ -38,7 +55,416 @@ COcclusion::AddOne(float x, float y, float z, float width, float length, float h
...
+	CVector coors;
+	float outW, outH;
+
+	if (!NumActiveOccluders || !CalcScreenCoors(pos, &coors, &outW, &outH))
+		return false;
+
+	float side = CVector(width, length, height).Magnitude() / 4.0f;
+	float area = Max(outW, outH) * side; 
+
+	CVector vec1, vec2;
```
Comment (aap) [inline src/render/Occlusion.cpp:421]: maybe call this min and max corner or something?

File: src/render/Occlusion.cpp [hunk 4]
```diff
@@ -38,7 +55,416 @@ COcclusion::AddOne(float x, float y, float z, float width, float length, float h
...
+
+	for (int i = 0; i < NumActiveOccluders; i++) {
+		if (coors.z - (side * 0.85f) <= aActiveOccluders[i].radius)
+			continue;
+
+		if (aActiveOccluders[i].IsPointWithinOcclusionArea(coors.x, coors.y, area))
+			return true;
+
+		if	(aActiveOccluders[i].IsPointWithinOcclusionArea(coors.x, coors.y, 0.0f) &&
+			(!CalcScreenCoors(vec1, &coors) || aActiveOccluders[i].IsPointWithinOcclusionArea(coors.x, coors.y, 0.0f)) &&
```
Comment (aap) [inline src/render/Occlusion.cpp:439]: like above, i think i'd do `if(CalcScreenCoors(...) && !IsPointWithinOcclusionArea(...)) continue;`

File: src/entities/Entity.cpp [hunk 5]
```diff
@@ -1157,3 +1158,38 @@ bool IsEntityPointerValid(CEntity* pEntity)
...
+		if (coors.z - (GetBoundRadius() * 0.85f) <= COcclusion::aActiveOccluders[i].radius) {
+			continue;
+		}
+
+		if (COcclusion::aActiveOccluders[i].IsPointWithinOcclusionArea(coors.x, coors.y, area)) {
+			return true;
+		}
+
+		if (COcclusion::aActiveOccluders[i].IsPointWithinOcclusionArea(coors.x, coors.y, 0.0f)) {
+			CVector v1 = m_matrix * CModelInfo::GetModelInfo(GetModelIndex())->GetColModel()->boundingBox.min;
```
Comment (aap) [inline src/entities/Entity.cpp:1181]: i think i'd call those min and max (also in AABB check below)

File: src/render/Occlusion.h [hunk 6]
```diff
@@ -1,12 +1,35 @@
...
+	float y1;
+	float x2;
+	float y2;
+	float length; //not sure
+};
+
+class CActiveOccluder {
+
+public:
+	//size = 0x80
```
Comment (aap) [inline src/render/Occlusion.h:14]: no need to keep track of sizes

File: src/render/Occlusion.h [hunk 7]
```diff
@@ -1,12 +1,35 @@
 #pragma once
 
+struct ActiveOccluderLine {
+	float x1;
+	float y1;
+	float x2;
+	float y2;
+	float length; //not sure
```
Comment (aap) [inline src/render/Occlusion.h:8]: it's correct

File: src/render/Occlusion.cpp [hunk 8]
```diff
@@ -38,7 +55,416 @@ COcclusion::AddOne(float x, float y, float z, float width, float length, float h
...
+COccluder::ProcessLineSegment(int a1, int a2, CActiveOccluder *occl) {
+	if (!gOccluderCoorsValid[a1] && !gOccluderCoorsValid[a2])
+		return false;
+
+	float x1, y1, x2, y2;
+
+	CVector vec1 = TheCamera.m_viewMatrix * gOccluderCoors[a1];
+	CVector vec2 = TheCamera.m_viewMatrix * gOccluderCoors[a2];
+
+	CVector coors1, coors2;
```
[the remark this answers is not in the document]
Comment (majesticCoding) [reply src/render/Occlusion.cpp:120]: how if I have to use them in CalcScreenCoors()?

**Pull request merged**